collector/lib/CollectorConfig.cpp

```diff
@@ -422,33 +425,120 @@ void CollectorConfig::YamlConfigToConfig(YAML::Node& yamlConfig) {
-    CLOG(FATAL) << "Failed to open the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to open the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::ParserException& e) {
-    CLOG(FATAL) << "Failed to parse the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to parse the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::Exception& e) {
-    CLOG(FATAL) << "An error occurred while loading the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "An error occurred while loading the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const std::exception& e) {
-    CLOG(FATAL) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    CLOG(ERROR) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    return false;
+  }
+
+  return true;
+}
+
+void CollectorConfig::WaitForFileToExist(const std::filesystem::path& filePath) {
+  int count = 0;
+  while (!std::filesystem::exists(filePath) && !thread_.should_stop()) {
+    sleep(1);
+    count++;
+    if (count > 45) {
+      std::unique_lock<std::shared_mutex> lock(mutex_);
+      runtime_config_.reset();
+    }
+  }
+}
+
+int CollectorConfig::WaitForInotifyAddWatch(int fd, const std::filesystem::path& filePath) {
+  while (!thread_.should_stop()) {
+    int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+    if (wd < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "Failed to add inotify watch for " << filePath;
+    } else {
+      return wd;
+    }
+    sleep(1);
   }
+
+  return -1;
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  int fd = inotify_init();
+  if (fd < 0) {
+    CLOG(ERROR) << "inotify_init() failed: " << StrError();
+    CLOG(ERROR) << "Runtime configuration will not be used.";
+    return;
+  }
+  WaitForFileToExist(filePath);
+  int wd = WaitForInotifyAddWatch(fd, filePath);
+  bool success = HandleConfig(filePath);
+  if (!success) {
+    CLOG(FATAL) << "Unable to parse configuration file: " << filePath;
+  }
+
+  char buffer[1024];
+  while (!thread_.should_stop()) {
+    int length = read(fd, buffer, sizeof(buffer));
+    if (length < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "Unable to read event for " << filePath;
+    }
+
+    struct inotify_event* event;
+    for (int i = 0; i < length; i += sizeof(struct inotify_event) + event->len) {
+      event = (struct inotify_event*)&buffer[i];
+      if (event->mask & IN_MODIFY) {
+        HandleConfig(filePath);
+      } else if ((event->mask & IN_MOVE_SELF) || (event->mask & IN_DELETE_SELF) || (event->mask & IN_IGNORED)) {
```

Comment:
The infinite loop you are getting is due to `IN_IGNORED` in this condition, from the man page:
```
           IN_IGNORED
                  Watch was removed explicitly (inotify_rm_watch(2)) or automatically (file was deleted, or filesystem  was
                  unmounted).  See also BUGS.
```

That means we are getting a move or delete event, then manually removing the watch, which triggers an IN_IGNORED event, we get that, go back in this place, remove the watch create it again get another IN_IGNORED event, etc...

Just remove `IN_IGNORED`, move or delete should be enough, we don't need to do anything in the ignore case.
```suggestion
      } else if ((event->mask & IN_MOVE_SELF) || (event->mask & IN_DELETE_SELF)) {
```